logical_type.go

```diff
@@ -304,6 +308,69 @@ func makeDecimalFixedCodec(st map[string]*Codec, enclosingNamespace string, sche
 	return c, nil
 }
 
+func makeValidatedStringCodec(st map[string]*Codec, enclosingNamespace string, schemaMap map[string]interface{}) (*Codec, error) {
+	patternFn, ok := schemaMap["pattern"]
+	if !ok {
+		return nil, errors.New("cannot create validated-string logical type without pattern")
+	}
+	if _, ok := schemaMap["name"]; !ok {
+		schemaMap["name"] = "string.validated-string"
+	}
+	c, err := registerNewCodec(st, schemaMap, enclosingNamespace)
+	if err != nil {
+		return nil, err
+	}
+
+	pattern := strings.TrimSpace(patternFn.(string))
+	if m[pattern] == nil {
+		m[pattern] = regexp.MustCompile(pattern)
+	}
+
+	c.binaryFromNative = validatedStringBinaryFromNative(c.binaryFromNative, pattern)
+	c.textualFromNative = validatedStringTextualFromNative(c.textualFromNative, pattern)
+	c.nativeFromBinary = validatedStringNativeFromBinary(c.nativeFromBinary, pattern)
+	c.nativeFromTextual = validatedStringNativeFromTextual(c.nativeFromTextual, pattern)
+	return c, nil
+}
+
+func validatedStringBinaryFromNative(fromNativeFn fromNativeFn, pattern string) fromNativeFn {
+	return func(b []byte, d interface{}) ([]byte, error) {
+		return stringBinaryFromNative(b, d)
+	}
+}
+
+func validatedStringTextualFromNative(fromNativeFn fromNativeFn, pattern string) fromNativeFn {
+	return func(b []byte, d interface{}) ([]byte, error) {
+		return stringTextualFromNative(b, d)
+	}
+}
+
+func validatedStringNativeFromBinary(fn toNativeFn, pattern string) toNativeFn {
+	return func(bytes []byte) (interface{}, []byte, error) {
+		fn, newBytes, _ := stringNativeFromBinary(bytes)
+
+		result := fn.(string)
+		if ok := m[pattern].MatchString(result); !ok {
+			return nil, bytes, fmt.Errorf("invalid string: %s, doesn't meet the pattern: %s", result, pattern)
+		}
+
+		return fn, newBytes, nil
+	}
+}
+
+func validatedStringNativeFromTextual(fn toNativeFn, pattern string) toNativeFn {
+	return func(bytes []byte) (interface{}, []byte, error) {
+		fn, newBytes, _ := stringNativeFromTextual(bytes)
+
+		result := fn.(string)
+		if ok := m[pattern].MatchString(result); !ok {
+			return nil, bytes, fmt.Errorf("invalid string: %s, doesn't meet the pattern: %s", result, pattern)
```

Comment:
Let's allow Go to quote the strings using the `%v` formatting verb, and eliminate the contraction.

`fmt.Errorf("cannot match input string against validation pattern: %q does not match %q", result, pattern)`